### PR TITLE
Key Vault Additions: Get Key and Sign

### PR DESCRIPTION
**Pull request by trevor-crypto**
I was using the javascript SDK and calling it from Rust, before I realized that this SDK exists. Thanks for saving me from a lot of mess!

For my use case, I really need these two API calls for Key Vault, so I went ahead and added them. I also ran into some problems with errors not being correctly parsed, so I also added a few and modified the `KeyVaultError` enum.  

I tried to model things off the existing code (`secret.rs`), but there were a few things that I was not sure about. For example, I made the datatypes equal to what I saw on the Azure SDK reference (linked below) and tried to name everything consistently and use the docs from there.

High level overview:
- Added Get Key https://docs.microsoft.com/en-us/rest/api/keyvault/getkey/getkey
- Added Sign https://docs.microsoft.com/en-us/rest/api/keyvault/sign/sign
- Added to and modified `KeyVaultError`
- Fixed various clippy lints (can remove if unnecessary) 

I am not sure of the status of this sdk, or if someone is already working on the changes I made, so feel free to close this if it is not needed.

Thanks

### CODE REVIEW

**Comment by trevor-crypto** [comment]
Will squash my commits once reviewed

**Review by MindFlavor — APPROVED** [review]
Seems fine to me! Excellent work!

**Comment by MindFlavor** [comment]
I would like to get an approval from @guywaldman as well before merging (he's the original author of this crate).

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:17]: In other languages, we call this `KeyVaultKey` and has a different shape, probably the closest semantically being Java: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKey.java

As new models are added, we should strive for consistency (though idiomatic to the language still takes precedent) rather than do it later.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:25]: `tags` and `managed` should hang off `KeyProperties`, which mainly wrap `KeyAttributes`.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:19]: `KeyAttributes` should be internal to the lib and wrapped by `KeyProperties`, which you can see here in Java: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyProperties.java#L29

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:34]: For consistency, `created_on` - at least as exposed on `KeyProperties` mentioned above.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:39]: `expires_on`

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:42]: `not_before`

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:54]: See https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/JsonWebKey.java for field naming.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:59]: Across our other languages we base64url-decode these, such that the field types are already optional buffers, e.g. `Option<Vec<u8>>`. Fields will have different sizes for different key types / key sizes.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:80]: `key_id`

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:77]: Ideally this is a union of known values (encrypt, decrypt, sign, verify, wrapKey, unwrapKey) or custom value in those cases a new value is supported. We let the service err accordingly. We call this `KeyOperation`.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:112]: Just `SignatureAlgorithm` like our other languages' SDKs.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:131]: The client should be named `KeyClient` per other languages' SDKs.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:141]: This must be optional. The service will use the latest.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:159]: We could probably accept this for now, but should open a bug. In all other languages' SDKs, we have a separate `CryptographyClient` that takes either a keyId URI or a JsonWebKey (the latter is new that is coming out soon) and performs cryptographic operations on that keyId. We don't do this in the `KeyClient` itself which is meant for key management.

Comment (heaths) [reply sdk/key_vault/src/key.rs:159]: Given the temporary nature, `key_name` and `key_version` could stay, but `value` should be `digest` and `alg` should be listed before `digest` and named `algorithm`. See https://github.com/Azure/azure-sdk-for-java/blob/08eadff95aa5a64dcefd8ef58a5bd29d536ab736/sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyClient.java#L320.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:165]: `KeyOperationResult` -> `SignResult`.

File: sdk/key_vault/src/sign.rs
Comment (heaths) [inline sdk/key_vault/src/sign.rs:14]: `SignResult`. The algorithm should also be passed back from the `sign` call for ease of passing back to `verify`, for example (when added). See https://github.com/Azure/azure-sdk-for-java/blob/08eadff95aa5a64dcefd8ef58a5bd29d536ab736/sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/models/SignResult.java#L13

**Review by heaths — CHANGES_REQUESTED** [review]
Thanks for adding this code, but it would be great to be consistent with our other languages SDKs while doing so than having to fix it all up later. See comments for links to examples (in Java, given use cases are more similar, though naming is snake_case instead of camelCase, of course).

File: sdk/key_vault/src/key.rs
Comment (trevor-crypto) [reply sdk/key_vault/src/key.rs:17]: Thanks for reviewing! It is good to be consistent, but it's a little strange to me why the actual REST API docs are not the one, final truth. I will make the changes to coincide with the Java SDK, per your request :slightly_smiling_face:

Comment (heaths) [reply sdk/key_vault/src/key.rs]: We are working to get to a place where the swagger and SDKs are more aligned. This is a much older swagger and names don't conform to normal languages' conventions (in many cases), and structures are sometimes changed based on user experience research. But across languages' SDKs, we try to be consistent when idiomatic differences are insignificant, e.g. Java, JavaScript, and Python might call something `listPropertiesOfSecrets` but .NET doesn't use "list" as a verb, instead using something like `GetPropertiesOfSecrets`.

General guidelines can be found here: https://azure.github.io/azure-sdk/general_introduction.html

Thank you for considering these changes!

File: sdk/identity/src/token_credentials/managed_identity_credentials.rs
Comment (guywaldman) [inline sdk/identity/src/token_credentials/managed_identity_credentials.rs:25]: Maybe I'm missin something, but is there a need for a closure here? We know the value ahead of time, no?

**Review by guywaldman — APPROVED** [review]

**Comment by guywaldman** [comment]
Hi, sorry for the delayed response - these changes look great IMO. Thank you for this awesome contribution.

I developed the original Key Vault crate and admittedly did not invest too much effort in aligning it to the other SDKs (this was a smaller-scale hobby project back then), which is why these are all welcome changes :)

In the upcoming weeks, assuming my capacity allows for it, I plan to improve this crate. So if there are other operations that are important for you feel free to open issues and I'll do my best to prioritize them for you.

File: sdk/identity/src/token_credentials/managed_identity_credentials.rs
Comment (rylev) [reply sdk/identity/src/token_credentials/managed_identity_credentials.rs:25]: This saves an allocation even the variable is set. How it was originally written we would allocate a string even if we never used it. Small optimization, but why not?

**Comment by rylev** [comment]
Ping to @heaths who I think should merge this assuming he thinks the changes since his review look good.

File: sdk/identity/src/token_credentials/managed_identity_credentials.rs
Comment (guywaldman) [reply sdk/identity/src/token_credentials/managed_identity_credentials.rs:25]: Ah right. Sure, LGTM.

Comment (trevor-crypto) [reply sdk/identity/src/token_credentials/managed_identity_credentials.rs:25]: Yeah, it is a `clippy` lint warning. I started going through them and then realized it wouldn't be the right PR to fix them all, so I stopped after a few.

**Comment by heaths** [comment]
I'm taking a look through this now. Sorry for the delay in getting back to it.

@guywaldman if you want to sync-up offline, that'd be great. I work with the KV and Managed HSM teams to develop our existing Azure SDKs, and currently working on the unofficial Azure SDK for Rust guidelines on our site here: https://azure.github.io/azure-sdk.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:285]: Algorithm should be listed first, though I wouldn't hold the PR up for this since I opened https://github.com/Azure/azure-sdk-for-rust/issues/197 to track moving this to `CryptographyClient`. That said, it'd be a minor change, I think, to do it now since some other small changes are needed.

**Review by heaths — CHANGES_REQUESTED** [review]
Just a couple minor things. I'll pay close attention to when this is updated and get it merged as soon as the changes made.

I made a couple comments that I wouldn't hold this PR for and will open bugs accordingly (once the PR comments are committed so I can reference them).

File: sdk/key_vault/examples/backup_secret.rs
Comment (heaths) [reply sdk/key_vault/examples/backup_secret.rs:21]: https://github.com/Azure/azure-sdk-for-rust/issues/198

File: sdk/key_vault/examples/backup_secret.rs
Comment (heaths) [reply sdk/key_vault/examples/backup_secret.rs:2]: https://github.com/Azure/azure-sdk-for-rust/issues/199

File: sdk/key_vault/src/client.rs
Comment (trevor-crypto) [reply sdk/key_vault/src/client.rs:78]: Is there a requirement to use https? This check will break the mock unit tests (mockito uses http, I guess)

File: sdk/key_vault/src/key.rs
Comment (trevor-crypto) [reply sdk/key_vault/src/key.rs:168]: When it is `None` serde will just ignore it and not send anything. It will not be included in the json

File: sdk/key_vault/src/key.rs
Comment (trevor-crypto) [reply sdk/key_vault/src/key.rs:205]: I removed it, but left it on `JsonWebKey` as an `Option<Vec<String>>`

File: sdk/identity/src/token_credentials/cli_credentials.rs
Comment (trevor-crypto) [reply sdk/identity/src/token_credentials/cli_credentials.rs:15]: Added tests to address this concern

**Comment by trevor-crypto** [comment]
@heaths thank you for the feedback. I tried to address everything without making too many giant changes. If everything looks ok, then I can squash my commits (unless you can do it through github).

File: sdk/key_vault/src/client.rs
Comment (heaths) [reply sdk/key_vault/src/client.rs]: We do in all the other SDKs, yes.

File: sdk/key_vault/src/client.rs
Comment (heaths) [inline sdk/key_vault/src/client.rs:38]: I'm assuming the scheme can be passed given you're calling `Url::parse()`. If not, that's concerning but could be addressed in a separate PR. Samples for creating it in other languages SDKs' are similar to:

```csharp
new KeyClient(new Uri("https://myvault.vault.azure.net")); // optional args elided
```

Granted, all the other languages take the URI as a string, but still with the scheme.

File: sdk/key_vault/src/client.rs
Comment (heaths) [inline sdk/key_vault/src/client.rs:62]: This will only work in the public cloud - not other environments. I don't think it's worth included at all. I'll open a separate issue to remove it. In general, we want to avoid anything that fixes a client to a single environment. Quite a few (including KV) have different domain suffixes in different clouds.

**Review by heaths — APPROVED** [review]

File: sdk/key_vault/src/client.rs
Comment (heaths) [reply sdk/key_vault/src/client.rs:62]: https://github.com/Azure/azure-sdk-for-rust/issues/201

**Comment by heaths** [comment]
Thank you, @trevor-crypto, for all your work in updating this! Most of this will result in fewer changes when we really start unifying the SDKs with the others we have (except where idiomatic differences take precedent).

File: sdk/identity/src/token_credentials/default_credentials.rs
Comment (cataggar) [inline sdk/identity/src/token_credentials/default_credentials.rs:17]: This effectively sets the `include_cli_credential`, `include_managed_identity_credential`, & `include_environment_credential` to `false`, which is but #526.

**Pull request merged**
